### PR TITLE
YAML files generation and merge

### LINKED ISSUE #47: Refactor YAML files
We want to separate the YAML configuration files in smaller YAML files where each of them will contain a separate part of the codebase:

- Robot parameters (number of modules, physics parameters, type of CAN bus ecc ...) for the two versions of the robot
- Digital twin parameters (xacro)
- Sensors (realsense and other sensors in future)
- Reseq controllers

You can find the config files in this directory: `reseq_ros2`/`reseq_ros2`/`config`

The idea is to create a script which will generate the YAML file for the specific configuration we will want to launch, including the other base YAML files. 

See #37 for more info

### PR DESCRIPTION
**Pull request by Yusuf-Serdar**
Closes #47

### CODE REVIEW

File: reseq_ros2/scripts/generate_configs.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+# Set the path to the config directory relative to the scripts directory
+config_path = os.path.join(os.path.dirname(__file__), '../config')
+temp_config_path = os.path.join(config_path, 'temp')
+
+# Ensure the temp directory exists
+os.makedirs(temp_config_path, exist_ok=True)
+
+# Function to merge multiple YAML files into a single dictionary
+def merge_yaml(file_paths):
```
Comment (lorixhack) [inline reseq_ros2/scripts/generate_configs.py:12]: This function can be removed

Comment (lorixhack) [inline reseq_ros2/scripts/generate_configs.py:12]: This function can be removed

File: reseq_ros2/scripts/generate_configs.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    # If include key is present, merge the included files
+    if 'include' in main_config:
+        include_files = main_config['include']
+        main_config = create_configs(main_config, include_files)
+        del main_config['include']  # Clean up the include key after merging
+    
+        # Save the final merged configuration to a temporary YAML file
+        temp_filename = f"{os.path.splitext(include_file)[0]}.yaml"
+        with open(os.path.join(temp_config_path, temp_filename), 'w') as outfile:
+            yaml.dump(main_config, outfile, default_flow_style=False)
```
Comment (lorixhack) [inline reseq_ros2/scripts/generate_configs.py:44]: The script shouldn't create the config file for all the configurations, it should only for that one we intend to launch

**Review by lorixhack — CHANGES_REQUESTED** [review]
Good job, it is a great start. I have only one objection about how the script works. I don't think that the script should create always all the config files (`reseq_mk1_can.yaml`, `reseq_mk1_vcan.yaml`, `reseq_mk1_vcan.yaml`); it would be better if the script creates the config file related to the argument we pass when we launch ROS.

For example, in case we don't pass any file as argument, the default is `reseq_mk1_can.yaml`. In this case, the script should merge and create only the YAML file related to the physical version of mk1.

File: reseq_ros2/config/reseq_mk1_can.yaml [hunk 3]
```diff
@@ -10,73 +16,4 @@ modules:
...
-  CylinderOnJetsonRadius: 0.026
-  CylinderOnJetsonShift: 0.012
-  CylinderRealsenseLength: 0.055
-  CylinderRealsenseRadius: 0.026
-  RealsenseLength: 0.026
-  RealsenseHeight: 0.02
-  RealsenseWidth: 0.09
-  RealsenseOriginSystem: 0.0162
\ No newline at end of file
+realsense_config: "realsense_rgb_motion.yaml"
```
Comment (luca-byte) [inline reseq_ros2/config/reseq_mk1_can.yaml:19]: This is the thing, the RealSense file name is passed and the launch file (obviously) cannot find it the temp folder, where it is looking for config files

**Review by luca-byte — CHANGES_REQUESTED** [review]
First and foremost, great job 💯 

This is exactly the design choice that we were looking for since it deals very well with the modularity of the configuration. Very well done!! There are just some problems in dealing with the peculiarities of the RealSense and especially the ROS2 controllers, but probably we defined the task with some imprecisions. I tried to clarify the special handling that they need.

I also noticed that the launch file was not working because you requested a launch argument before declaring it, but i suggested a solution that in my tests work

In the end I also listed some minor issues, which we can easily correct to avoid that they become bigger problems

File: reseq_ros2/scripts/generate_configs.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        os.makedirs(temp_path, exist_ok=True)
+
+# Function to merge multiple YAML files into a single dictionary
+def merge_yaml(file_paths):
+    merged_data = {}
+    for file_path in file_paths:
+        with open(file_path, 'r') as file:
+            data = yaml.safe_load(file)
+            merged_data.update(data)
+    return merged_data
```
Comment (luca-byte) [inline reseq_ros2/scripts/generate_configs.py:26]: I do not understand what this function is for and when it is used

File: reseq_ros2/launch/reseq_launch.py [hunk 5]
```diff
@@ -177,7 +180,28 @@ def launch_setup(context, *args, **kwargs):
...
+            'config_file',
+            default_value=default_filename,
+        ),
+        generate_configs,
+        # Wait for the config generation process to complete before proceeding
+        RegisterEventHandler(
+            OnProcessExit(
+                target_action=generate_configs,
+                on_exit=[LogInfo(msg="Configuration files generated."),
+                         OpaqueFunction(function=launch_setup)
```
Comment (luca-byte) [inline reseq_ros2/launch/reseq_launch.py:202]: This prints the message even if the script had an exception:
```
[INFO] [generate_configs-1]: process started with pid [2529]
[generate_configs-1] python3: can't open file '.../ros2_ws/install/reseq_ros2/share/reseq_ros2/scripts/generate_configs.py': [Errno 2] No such file or directory
[ERROR] [generate_configs-1]: process has died [pid 2529, exit code 2, cmd 'python3 .../ros2_ws/install/reseq_ros2/share/reseq_ros2/scripts/generate_configs.py reseq_mk1_can.yaml'].
[INFO] [launch.user]: Configuration files generated.
[ERROR] [launch]: Caught exception in launch (see debug for traceback): [Errno 2] No such file or directory: '.../ros2_ws/install/reseq_ros2/share/reseq_ros2/config/temp/reseq_mk1_can.yaml'
```

**Review by luca-byte — COMMENTED** [review]
Great job!! 💯 

Just some smaller improvements and then I think that we can merge it!

File: reseq_ros2/scripts/generate_configs.py (hunk 4 above)
Comment (Yusuf-Serdar) [reply reseq_ros2/scripts/generate_configs.py:26]: Yes, you are right. It was probably used in the previous version of the script and I have forgotten to get rid of it.

**Review by luca-byte — APPROVED** [review]
LGTM

**Review by lorixhack — APPROVED** [review]
LGTM

**Pull request merged**